### ISSUE TITLE
Logic cotrolling devices best practicies 

### ISSUE THREAD

**Issue by holotrack** [kubeedge/kubeedge#5340]
Hello. I need to implement logic and dashboard which will be automatically controlling my devices based on settings. How it should be handled? Just by updating yaml files from container with kubectl or there is some better option for it (api?, grpc?)?

Example:

I have temperature sensor and smart switch which will be turning on/off fun if temperature exceed some level or fall down below it. Manually i will just update "write" field in CRD device yaml to turn on/off fun but i don't want to look on it and switching it manually by editing yaml.

How to automate this process with kubeedge, how to change properly statuses of devices? Im not seeing anywhere example of such real world solution for kubeedge. All docs which i found and videos showing only how to change it manually by editing RBACs

**Comment by wbc6080** [comment]
You can implement an application to monitor the status reported by the device and update the device according to the status. You can refer to the specific implementation in [e2e test](https://github.com/kubeedge/kubeedge/blob/master/tests/e2e/device/device.go#L106) for how to update the device by edgeclient.

**Comment by holotrack** [comment]
> You can implement an application to monitor the status reported by the device and update the device according to the status.

And how to do that? Is there some simple example like for mappers on which someone without good knowledge of Go can base and implement it in any other language?

> You can refer to the specific implementation in [e2e test](https://github.com/kubeedge/kubeedge/blob/master/tests/e2e/device/device.go#L106) for how to update the device by edgeclient.

According to my understanding your library is using [client-go](https://github.com/kubeedge/kubeedge/blob/master/pkg/client/clientset/versioned/typed/devices/v1beta1/devices_client.go#L26) to contact kubernetes api to update statuses of devices directly in yaml? Am i right?

Maybe im missing something but eg on your trafic light example:
https://github.com/kubeedge/examples/blob/master/traffic-light/message.go

Lights are updating to state of kubernetes yaml, but its not like lights should work. They should change for some time from one to another. Where and with usage of what (some rest api, grpc, anything else of core controller?) should be implemented such logic?

Basically my question is how real world application should be designed on top of kubeedge to make them useful. Im seeing right now only tool for showing statuses of devices but i want to use those statuses to make some actions on "active" devices, like eg automatically turn on fan or turn off light (can be traffic lights).

**Comment by wbc6080** [comment]
Since Go users account for the majority in the community, we temporarily provide a mapper based on the Go language to manage devices. Mappers based on multiple languages ​​are still being planned and implemented. In kubeedge, the way we manage devices is similar to pods in k8s, mainly macroscopic create, delete, update. However, it is still incomplete functionality to directly open and close physical devices. At the current stage, I think you can use mapper's data push capability to push the collected device data to your own built application, and your application will turn the device on and off based on the collected device data. At the same time, I will also record your needs, discuss them in the community and plan for subsequent versions.

**Comment by holotrack** [comment]
> Since Go users account for the majority in the community, we temporarily provide a mapper based on the Go language to manage devices. Mappers based on multiple languages ​​are still being planned and implemented. In kubeedge, the way we manage devices is similar to pods in k8s, mainly macroscopic create, delete, update. However, it is still incomplete functionality to directly open and close physical devices. At the current stage, I think you can use mapper's data push capability to push the collected device data to your own built application, and your application will turn the device on and off based on the collected device data. At the same time, I will also record your needs, discuss them in the community and plan for subsequent versions.

Thank you for your reply. So for now changing devices statuses is only by updating devices "yamls" with eg go-client  or just by use kubectl directly in my app?

Or i should change device status directly from my app and mapper will update status to kubeedge/mqtt?

If i understand it correctly do you already have some plan how handle that "triggered" device statuses changes? You want to create some additional API for it or some other ideas?

**Comment by holotrack** [comment]
So what i understand for now i should based on mqtt/kubernetes data directly contact device and set it up and then mapper will update status in kubeedge CRD or just directly update CRD with eg client-go or with kubectl itself?

Do you have any idea how it will be looks like in future some additional service with API or something? Do you have some other ideas for it? If not maybe good to consider some gRPC or Cap'n proto service? Thanks to it you can create .proto file and based on it generate client libraries for all languages supported by it?

**Comment by holotrack** [comment]
> Since Go users account for the majority in the community, we temporarily provide a mapper based on the Go language to manage devices. Mappers based on multiple languages ​​are still being planned and implemented. In kubeedge, the way we manage devices is similar to pods in k8s, mainly macroscopic create, delete, update. However, it is still incomplete functionality to directly open and close physical devices. At the current stage, I think you can use mapper's data push capability to push the collected device data to your own built application, and your application will turn the device on and off based on the collected device data. At the same time, I will also record your needs, discuss them in the community and plan for subsequent versions.

Is it what i was asking for? https://github.com/kubeedge/mappers-go/blob/main/mappers/virtualdevice-sdk/README.md#restful-api-usage-example

**Comment by wbc6080** [comment]
The example shown here was before version 1.12, using mqtt to transmit messages to update the device. However, in version 1.12 we proposed `DMI` to manage the device uniformly in the form of k8s resources. This part has been deprecated. If you want to know more about `DMI` and `mapper`, you can refer to [DMI](https://deploy-preview-439--kubeedge.netlify.app/docs/developer/dmi) and [mapper](https://deploy-preview-439--kubeedge.netlify.app/docs/developer/mappers/)

**Comment by wbc6080** [comment]
Also here are the changelogs about [DMI](https://github.com/kubeedge/kubeedge/blob/master/CHANGELOG/CHANGELOG-1.12.md#introducing-alpha-implementation-of-next-gen-cloud-native-device-management-interfacedmi) in version 1.12 and about mapper's automatic generation framework [mapper-framework](https://github.com/kubeedge/kubeedge/blob/master/CHANGELOG/CHANGELOG-1.15.md#support-alpha-version-of-dmi-dateplane-and-mapper-framework) in version 1.15.

**Comment by holotrack** [comment]
@wbc6080 Thank for answere and links. Its all what i wanted to know. Closing issue.
